File: Android/Week2/app/src/test/java/cs/mad/week2lab/Week2LabTests.kt
package cs.mad.week2lab

import cs.mad.week2lab.entities.Flashcard
import cs.mad.week2lab.entities.FlashcardSet
import org.junit.Test
import kotlin.reflect.KProperty

class Week2LabTests {

    fun getFlashcardSets(): Array<FlashcardSet> {
        return arrayOf(
            FlashcardSet("one"),
            FlashcardSet("two"),
            FlashcardSet("three"),
            FlashcardSet("four"),
            FlashcardSet("five"),
            FlashcardSet("six"),
            FlashcardSet("seven"),
            FlashcardSet("eight"),
            FlashcardSet("nine"),
            FlashcardSet("ten")
        )
    }

    fun getFlashcards(): Array<Flashcard> {
        return arrayOf(
            Flashcard("one", "the first number"),
            Flashcard("two", "the second number"),
            Flashcard("three", "the third number"),
            Flashcard("four", "the fourth number"),
            Flashcard("five", "the fifth number"),
            Flashcard("six", "the sixth number"),
            Flashcard("seven", "the seventh number"),
            Flashcard("eight", "the eighth number"),
            Flashcard("nine", "the ninth number"),
            Flashcard("ten", "the tenth number")
        )
    }


    @Test
    fun testFlashcard() {
        assert(sizeAndType<Flashcard>(getFlashcards()))

        assert(Flashcard::term is KProperty<String>)
        assert(Flashcard::definition is KProperty<String>)
    }

    @Test
    fun testFlashcardSet() {
        assert(sizeAndType<FlashcardSet>(getFlashcardSets()))
        assert(FlashcardSet::title is KProperty<String>)
    }

    private fun <T> sizeAndType(array: Array<T>) = sizeAndType(array.asList())
    private fun <T> sizeAndType(collection: Collection<T>) = collection.size == 10
    private fun <T> sizeAndType(other: Any) = false
}

File: iOS/Week2Lab-iOS/Week2LabTests/Week2LabTests.swift
//
//  Week2LabTests.swift
//  Week2LabTests
//
//  Created by Andrew Taylor on 1/15/23.
//

import XCTest
@testable import Week2Lab

final class Week2LabTests: XCTestCase {
    func getFlashcards() -> Array<Flashcard> {
        return [
            Flashcard(term: "one", definition: "the first number"),
            Flashcard(term: "two", definition: "the second number"),
            Flashcard(term: "three", definition: "the third number"),
            Flashcard(term: "four", definition: "the fourth number"),
            Flashcard(term: "five", definition: "the fifth number"),
            Flashcard(term: "six", definition: "the sixth number"),
            Flashcard(term: "seven", definition: "the seventh number"),
            Flashcard(term: "eight", definition: "the eighth number"),
            Flashcard(term: "nine", definition: "the ninth number"),
            Flashcard(term: "ten", definition: "the tenth number")
        ]
    }
    
    func getFlashcardSets() -> Array<FlashcardSet> {
        return [FlashcardSet(title: "one"), FlashcardSet(title: "two"), FlashcardSet(title: "three"), FlashcardSet(title: "four"), FlashcardSet(title: "five"), FlashcardSet(title: "six"), FlashcardSet(title: "seven"), FlashcardSet(title: "eight"), FlashcardSet(title: "nine"), FlashcardSet(title: "ten")]
    }
    
    func testFlashcard() throws {
        let flashcards = getFlashcards()
        assert(flashcards.count == 10)
        assert(flashcards is Array<Flashcard>)
        
        let mirror = Mirror(reflecting: flashcards[0])
        assert(mirror.children.count == 2)
        assert(
            mirror.children.contains { (label: String?, _: Any) in
                label == "term"
            }
        )
        assert(
            mirror.children.contains { (label: String?, _: Any) in
                label == "definition"
            }
        )
    }

    func testFlashcardSet() throws {
        let flashcardSets = getFlashcardSets()
        assert(flashcardSets.count == 10)
        assert(flashcardSets is Array<FlashcardSet>)
        
        let mirror = Mirror(reflecting: flashcardSets[0])
        assert(mirror.children.count == 1)
        assert(
            mirror.children.contains { (label: String?, _: Any) in
                label == "title"
            }
        )
    }
}
